pulp_container/app/registry_api.py

```diff
@@ -475,6 +475,144 @@ def get_queryset(self, *args, **kwargs):
+
+    def get(self, request):
+        bad = False
+        req_repository = None
+        req_tags = set()
+        req_os = None
+        req_architecture = None
+        req_label_exists = {"org.flatpak.ref"}
+        req_label_values = {}
+        for (key, values) in request.query_params.lists():
+            if key == "repository":
+                req_repository = values[0]
+                if any(v != req_repository for v in values[1:]):
+                    bad = True
+                    break
+            elif key == "tag":
+                req_tags.update(values)
+            elif key == "os":
+                req_os = values[0]
+                if any(v != req_os for v in values[1:]):
+                    bad = True
+                    break
+            elif key == "architecture":
+                req_architecture = values[0]
+                if any(v != req_architecture for v in values[1:]):
+                    bad = True
+                    break
+            elif key.startswith("annotation:"):
+                # This implementation never announces any Annotations:
+                bad = True
+                break
+            elif key.startswith("label:"):
+                if key.endswith(":exists"):
+                    if any(v != "1" for v in values):
+                        bad = True
+                        break
+                    label = key[len("label:") : len(key) - len(":exists")]
+                    req_label_exists.add(label)
+                else:
+                    if any(v != values[0] for v in values[1:]):
+                        bad = True
+                        break
+                    label = key[len("label:") :]
+                    req_label_values[label] = values[0]
+            else:
+                bad = True
+                break
+
+        results = []
+
+        if not bad:
+            # This part copied from CatalogView:
+            queryset = models.ContainerDistribution.objects.all().only("base_path")
+            distribution_permission = "container.pull_containerdistribution"
+            namespace_permission = "container.namespace_pull_containerdistribution"
+            public_repositories = queryset.filter(private=False)
+            repositories_by_distribution = get_objects_for_user(
+                self.request.user, distribution_permission, queryset
+            )
+            namespaces = models.ContainerNamespace.objects.all()
+            repositories_by_namespace = get_objects_for_user(
+                self.request.user, namespace_permission, namespaces
+            )
+            repositories_by_namespace = queryset.filter(namespace__in=repositories_by_namespace)
+            accessible_repositories = repositories_by_distribution & repositories_by_namespace
+            repositories = (public_repositories | accessible_repositories).distinct()
+
+            if req_repository is not None:
+                repositories = repositories.filter(base_path=req_repository)
```

Comment:
No, that `req_repository` models what other languages would model with something like Haskell `Maybe String`, either `Nothing` (i.e., `None` here) or `Just s` (i.e., a string object here).

Reply:
Shouldn't `repositories` always return a query of 1 (or 0). `base_path` is a unique field for distributions so filtering by it will return a unique distribution.